### DISCUSSION TITLE
Non-interactive usage of SDM: options for passwords and Y/N answers

### DISCUSSION THREAD

**Discussion by securechicken** [gitbls/sdm#8] [Ideas]
Dear SDM community,

I recently tested SDM, it's an awesome tool.
I however wanted to test various non-interactive scenarios, where SDM would be called from scripts/tools, and would require NO user interaction.
I found it is impossible to fully use SDM non-interactively as it is now: 
- SDM asks for a terminal Y/N read if a --customize image is customized again;
- SDM asks for a terminal Y/N read to determine if users passwords should be set;
- SDM asks for at least one password from terminal.

I tried to input data by piping echo/cat commands to SDM, or by redirecting a file to it, but this hurts systemd-nspawn calls then, as a TTY is supposed to be passed as their input (i.e `< $(tty)`)

I think it would be interesting to enable fully non-interactive SDM use scenarios, by dropping the `$(tty)` input use, and/or by adding command-line options to answers Y/N questions (e.g. --image-reuse y|n, --same-password y|n, --root-password y|n), as well as to set users password(s) programatically (e.g. --set-user-password).

Thanks for your consideration and opinions!

**Comment by gitbls** [comment]
Great suggestions! I should be able to get to these in the next week or two. Stay tuned!

**Reply by securechicken** [reply]
Glad you agree!
For the record, I still managed to do what I want thanks to the fact most tools are plain Bash (and are readable as so), but it required to "patch" sdm-phase1 and sdm scripts to replace some `askyn` and `getpasswd` calls according to my test needs; I think this is not portable at all :)
I would have attempted a PR, but I sincerely believe it's better if specifications and use cases are defined by you as the designer - I could however implement (parts of) your definition/specs in my own time, if needed.

**Reply by gitbls** [reply]
I appreciate the offer, but these are all fairly straightforward. It's simply a matter of having the time to have at it, but unfortunately I'm tied up for the next several days. Once I free up this is the first item on my todo list.

**Comment by gitbls** [comment]
Had a chance to look at this in a bit more detail before I go busy for a couple of days. Not sure what --root-password y/n would be used for, since --rootpwd is already there. Here's my current thinking on new switches to meet your needs. LMK your thoughts:

--image-reuse y/n  # Skip the re-customize question if "y" (may change to --recustomize, still thinking about this)
--same-password y/n # Skip the "use the same password" question if "y"
--set-root-password password # Set the root password if specified (and --rootpwd is set)
--set-pi-password password  # Set the user pi password if specified
--set-user-password password # Set the password for the "--user username" account if specified

If "--same-password y", then the password for pi/created-user will all be set to the password for "--set-user-password" value (if specified, or "--set-pi-password" if specified. (root password will only be set in this case if --rootpwd is set).

I'll also look into adding some of this to the "burn" code, since it might make sense to change passwords when the image is burned, for instance in the case where SD Cards are sent to multiple, independent sites. Thoughts?

This may all change when I update the code, of course 🤣

**Reply by securechicken** [reply]
> Not sure what --root-password y/n would be used for, since --rootpwd is already there. 

My bad, I actually thought about this a few days ago, and did not remember the --rootpwd did it all already while opening this discussion.

> --image-reuse y/n # Skip the re-customize question if "y" (may change to --recustomize, still thinking about this)
> --same-password y/n # Skip the "use the same password" question if "y"
> --set-root-password password # Set the root password if specified (and --rootpwd is set)
> --set-pi-password password # Set the user pi password if specified
> --set-user-password password # Set the password for the "--user username" account if specified
> 
> If "--same-password y", then the password for pi/created-user will all be set to the password for "--set-user-password" value (if specified, or "--set-pi-password" if specified. (root password will only be set in this case if --rootpwd is set).

Seems perfect to solve the use cases I tested and to open for plenty more.

> I'll also look into adding some of this to the "burn" code, since it might make sense to change passwords when the image is burned, for instance in the case where SD Cards are sent to multiple, independent sites. Thoughts?

Make sense. I do passwords/users changes with customize because I just want non-RaspiOS defaults for a factory image; but it would probably be convenient to be able to change passwords while burning. However, I think this is trickier than it seems: I deleted the "pi" user at customize with a custom phase script for instance... so any pi password change at burn would not go well.

**Comment by gitbls** [comment]
Nearly done with the update. Here's what I landed on:
- --redo-customize -- Skip the prompt for redoing the customization if already done. 
- --password-pi _password_ -- Set the password for user Pi
- --password-user _password_ -- Ditto for the --user user
- --password-root _password_ -- Ditto for root
- --password-same y|n -- If "y", use the same password for all 3. If "n", don't ask if all 3 passwords should be the same. Password choice order: Password for --password-pi, if "" then --password-user, if "" then --password-root. If all are blank, then prompt user.

Using --batch prevents dropping into bash in the customized image.  Complete non-interactive operation now works. 

Passwords can be changed on the burn command, so making customized SD Cards for multiple sittes (or whatever) is now simple

Oh, and go ahead and delete that Pi user. 🤣  If you use --password-pi on the burn command and the user isn't there, you'll be notified, since you asked for it to be done.

Still need to do more testing, but coming soon!

**Comment by gitbls** [comment]
Changes are all checked in. I'd appreciate your feedback after you've had an opportunity to give it a try.

**Reply by securechicken** [reply]
I just tried a combination of `--redo-customize "y"`, `--password-user "${OBVIOUSLY_EASY_TO_GUESS}"`,  and `--password-same "y"`, without `--rootpwd`, and it works like a charm for a previous scenario I tested where I had no choice but to temporarily patch sdm tools. Awesome, thanks for this very pretty addition that now make the tool a platinum script friend :)
Must also add the CLI documentation is very clear on the way they are used.

**Reply by gitbls** [reply]
Great...And thanks for your excellent suggested improvement!
